### PR TITLE
W25/trinity/pet list endpoint

### PR DESCRIPTION
**Pull request by trinity-y**
## Notion ticket link

[Create Endpoint for Petlist](https://www.notion.so/uwblueprintexecs/Sprint-Board-4f2ffd2639ac41c4ad0f64818d84d5fa?p=1a010f3fb1dc80439b13fd5781faa1ac&pm=s)

## Implementation description
* for a given user, return a list of pet data (PetListItemDTO):
```
export interface PetListItemDTO {
  id: number;
  name: string;
  color: ColorLevel;
  taskCategories: Category[];
  status: PetStatus;
  lastCaredFor: string; // will hold a timestamp or 'One or more days ago'
  allTasksAssigned: boolean | null; // null if there are no tasks
  isAssignedToMe: boolean;
  photo?: string;
}
```
* this list returns all pets assigned to the current user first, and then all other pets. within those categories, the pets are sorted by urgency of care.
* node visibility based on user role is not implemented in the backend

## Product assumptions
* for pets with no tasks assigned, lastCaredFor is set to 'One or more days ago'
* sorting in terms of time/urgency was implemented by showing Needs Care, Occupied, then Does Not Need Care
    * did this instead of sorting based on start time since start times are lost after pet list is compiled + lots of pets have none or many start times

## Steps to test
1. fill activities, activity_type, pet, and user tables with test data
2. call pet list endpoint at `/pets/list/userId`

## What should reviewers focus on?
* are product assumptions  (listed above) correct
* I had to update typescript/ts-node so that I could use luxon (for date/time stuff and dealing with time zones) so make sure nothing broke 

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/implementations/petService.ts [hunk 1]
```diff
@@ -314,6 +316,207 @@ class PetService implements IPetService {
...
+      const sortedNotAssignedPetList = this.sortPetItemListByStatus(
+        notAssignedToUser,
+      );
+      const sortedPetList = sortedAssignedPetList.concat(
+        sortedNotAssignedPetList,
+      );
+      return sortedPetList;
+    } catch (error: unknown) {
+      Logger.error(getErrorMessage(error));
+      return [];
```
Comment (cindyhjli) [inline backend/typescript/services/implementations/petService.ts:517]: throw the error?

File: backend/typescript/services/implementations/petService.ts [hunk 2]
```diff
@@ -314,6 +316,207 @@ class PetService implements IPetService {
...
+  timeStringToDate(timeStr: string): DateTime {
+    const currentTime = DateTime.now().setZone(TIME_ZONE);
+    const fullDateString = `${currentTime.toFormat("DDDD")} ${timeStr}`;
+    return DateTime.fromFormat(fullDateString, "DDDD t");
+  }
+
+  colorLevelToColor(colorLevel: number): ColorLevel {
+    // (values are in descending order)
+    return Object.values(ColorLevel)[5 - colorLevel];
+  }
```
Comment (cindyhjli) [inline backend/typescript/services/implementations/petService.ts:334]: breaks if enum order is changed

File: backend/typescript/services/implementations/petService.ts [hunk 3]
```diff
@@ -314,6 +316,207 @@ class PetService implements IPetService {
...
+
+  // petList helper functions
+  dateToTimeString(date: DateTime): string {
+    return date.setZone(TIME_ZONE).toFormat("t");
+  }
+
+  timeStringToDate(timeStr: string): DateTime {
+    const currentTime = DateTime.now().setZone(TIME_ZONE);
+    const fullDateString = `${currentTime.toFormat("DDDD")} ${timeStr}`;
+    return DateTime.fromFormat(fullDateString, "DDDD t");
```
Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: Error handling needed for this line (invalid timeString possible)

Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: error handling for line (timeStr could be invalid)

File: backend/typescript/services/implementations/petService.ts [hunk 4]
```diff
@@ -314,6 +316,207 @@ class PetService implements IPetService {
...
+  }
+
+  colorLevelToColor(colorLevel: number): ColorLevel {
```
Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: rename to colorLevelToEnum

File: backend/typescript/services/implementations/petService.ts [hunk 5]
```diff
@@ -314,6 +316,207 @@ class PetService implements IPetService {
...
+            status: petActivity.status,
+            lastCaredFor: ONE_OR_MORE_DAYS_AGO,
+            allTasksAssigned: null, // null if there are no tasks
+            isAssignedToMe: false,
+          };
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+        const petData = petIdToPetListItem[petActivity.pet_id];
+        // if the pet already exists in the hashMap
```
Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: clean up if nesting + comments for clarity

File: backend/typescript/services/implementations/petService.ts [hunk 6]
```diff
@@ -314,6 +320,218 @@ class PetService implements IPetService {
     }
     return id;
   }
+
+  // petList helper functions
```
Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: Move petList helper functions into typescript/utilities folder. 

Create new file typescript/utilities/common.ts to hold these helpers:
- dateToTimeString
- timeStringToDate
- colorLevelToEnum

File: backend/typescript/services/implementations/petService.ts
Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: To be renamed to PetTask

Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: Sorting implementation does not need to be defined in this already too long function, create a private method in PetService to handle this. Expect this to receive an unsortedPetList and to output a sortedPetList based on the sorting logic.

File: backend/typescript/services/implementations/petService.ts [hunk 7]
```diff
@@ -314,6 +325,193 @@ class PetService implements IPetService {
...
+              // Update assignment status
+              if (!petTask.user_id) {
+                petData.allTasksAssigned = false;
+              } else if (petTask.user_id === currUserId) {
+                petData.isAssignedToMe = true;
+              }
+            } else {
+              // Task has finished, update lastCaredFor time
+              const endTime = DateTime.fromJSDate(petTask.end_time);
+              if (petData.lastCaredFor === ONE_OR_MORE_DAYS_AGO) {
```
Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: this if block includes unnecessary and unclear nesting + doesn't handle all cases

File: backend/typescript/services/implementations/petService.ts [hunk 8]
```diff
@@ -314,6 +325,193 @@ class PetService implements IPetService {
...
+                if (endTime > beginningOfToday) {
+                  petData.lastCaredFor = dateToTimeString(endTime);
+                }
+              } else if (
+                petData.lastCaredFor &&
```
Comment (sthuray) [inline backend/typescript/services/implementations/petService.ts]: petData.lastCaredFor never set?

**Review by sthuray — CHANGES_REQUESTED** [review]
I've left a few comments but I'm sure there are still a few logic errors left since the last change.

I expect to see re-done screenshots of testing after every re-request for review. In the next batch of testing, I need to see extensive testing of every edge case, a screenshot example of that edge case, and a clear explanation for how the screenshot aligns with expected behaviour (can't merge this code in without knowing that edge cases were tested)

I still want to see thorough documentation of the if nesting block (for your understanding and mine) as well as an attempt to flatten nesting structure for clarity. This will also help you write testing for edge cases. 

edit 1: oops I misread the end_time, I think basic functionality is  still up and running

edit 2: nvm, it's still broken for several edge cases :(

**Review by he-is-harry — APPROVED** [review]
Thanks for updating the PR, code looks good to me!

**Review by sthuray — CHANGES_REQUESTED** [review]
Updated code with several changes:
- Corrected some logical errors
    - there was a lexicographical comparison between dateToString(endtime) < lastCaredFor rather than a time based comparison
    - some other small things
- Flattened conditionals for clarity/maintainability + added comments
- Added typing support for LastCaredFor

Testing is not proof-read, has several logical inconsistencies and errors throughout. On a quick look:
- it says Tiger has 2 tasks on the day (1007, 1008)
- The start/end time for task 1008 happens within the start/end time for 1007 (impossible)
- it says Tiger has pet status "occupied", but it says "needs care" in the pets table
- no task categories added for tiger despite being occupied (currently occupied icon should still be added)
- Kiwi: "ongoing unassigned task" (impossible)
- Rocky: status is "Needs Care" even though all tasks are completed
- No testing for multiple of the same task categories
etc. etc. 

Please proof-read testing and re-write the new testing. 
Additionally, include SQL queries for reviewers to try this testing themselves. + tell them how to adjust the time stamp for their own testing (ie. what time did you run the testing compared to the time stamps, is it before or after the tasks have ended)
